### ISSUE TITLE
Should support ETCD_ENDPOINTS (in additional to ETCD_AUTHORITY)

### ISSUE THREAD

**Issue by tomdee** [projectcalico/calico-containers#845]
Other projects (Kubernetes, Flannel) support an etcd configuration as a comma separated list of `{protocol}://{ip}:{port}`. Calico should support this too.

This relies on a felix change too (https://github.com/projectcalico/calico/issues/988)

I expect that ETCD_AUTHORITY would continue to be supported but if this new variable is set then that would take precedence.

**Comment by TrimBiggs** [comment]
This is related to #818, and a fix for this will fix #818 as well.

**Comment by fasaxc** [comment]
FYI, I tried out passing multiple hosts/ports to python-etcd in an effort to give a migration path from the old etcd port to the IANA-assigned one.  I found that, with `use_proxies=True`, which made sense for my use-case, I hit this issue:
```
>>> c = etcd.Client((("localhost", 1234), ("localhost", 4001)), allow_reconnect=True,use_proxies=True)
>>> c.get("calico")
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    c.get("calico")
  File "/home/gulfstream/git/calico/env27/local/lib/python2.7/site-packages/etcd/client.py", line 624, in ge
t
    return self.read(key)
  File "/home/gulfstream/git/calico/env27/local/lib/python2.7/site-packages/etcd/client.py", line 481, in re
ad
    timeout=timeout)
  File "/home/gulfstream/git/calico/env27/local/lib/python2.7/site-packages/etcd/client.py", line 798, in ap
i_execute
    self._machines_cache.remove(self._base_uri)
ValueError: list.remove(x): x not in list
```
Without setting `use_proxies`, it seems to work:
```
>>> c = etcd.Client((("localhost", 1234), ("localhost", 4001)), allow_reconnect=True)
>>> c.get("calico")
<class 'etcd.EtcdResult'>({'newKey': False, 'raft_index': 3116078, '_children': [{u'createdIndex': 9, u'modi
fiedIndex': 9, u'dir': True, u'key': u'/calico/bgp'}, {u'createdIndex': 22, u'modifiedIndex': 22, u'dir': Tr
ue, u'key': u'/calico/felix'}, {u'createdIndex': 7, u'modifiedIndex': 7, u'dir': True, u'key': u'/calico/ipa
m'}, {u'createdIndex': 4, u'modifiedIndex': 4, u'dir': True, u'key': u'/calico/v1'}], 'createdIndex': 4, 'mo
difiedIndex': 4, 'value': None, 'etcd_index': 83314, 'expiration': None, 'key': u'/calico', 'ttl': None, 'ac
tion': u'get', 'dir': True})
>>> c = etcd.Client((("localhost", 4001), ("localhost", 1234)), allow_reconnect=True)
>>> c.get("calico")
<class 'etcd.EtcdResult'>({'newKey': False, 'raft_index': 3116097, '_children': [{u'createdIndex': 7, u'modi
fiedIndex': 7, u'dir': True, u'key': u'/calico/ipam'}, {u'createdIndex': 4, u'modifiedIndex': 4, u'dir': Tru
e, u'key': u'/calico/v1'}, {u'createdIndex': 9, u'modifiedIndex': 9, u'dir': True, u'key': u'/calico/bgp'}, 
{u'createdIndex': 22, u'modifiedIndex': 22, u'dir': True, u'key': u'/calico/felix'}], 'createdIndex': 4, 'mo
difiedIndex': 4, 'value': None, 'etcd_index': 83314, 'expiration': None, 'key': u'/calico', 'ttl': None, 'ac
tion': u'get', 'dir': True})
>>> 
```

**Comment by fasaxc** [comment]
Note: to enable multiple host support, you have to set `allow_reconnects=True` and I think that breaks proxy support, because it triggers python-etcd to query etcd's cluster topology and use that for the reconnects.

When we are connected to a proxy only, I think we're supposed to set `use_proxies=True`, which is a flag they added after we started using python-etcd.

**Comment by matthewdupre** [comment]
Your test was on >0.4.1 right?  That added "* Allow the use of etcd proxies with reconnections"

**Comment by matthewdupre** [comment]
From an inspection it looks like it's still there (bit hard to be sure without running it though).

If _allow_reconnect, _wrap_request() calls _next_server() when it fails to connect.  That pops the next server out of _machines_cache (which is initially just the configured list).  Then because some_request_failed and _use_proxies, we try to remove the already popped server.

**Comment by matthewdupre** [comment]
One note on this one.  When we enable passing multiple endpoints to python-etcd we'll need to turn on allow_reconnect.  At this point we'll get nasty behaviour with proxies (or turn on use_proxies and hit the bug mentioned above).

The problem is with allow_reconnect, python-etcd will grab the addresses of the cluster nodes and consider connecting to them directly.  This isn't intended.

For now I suggest setting allow_reconnect only when passing multiple endpoints through.  This will restrict bad behaviour to the somewhat degenerate case of passing multiple proxy addresses in, and still allow the case of one local proxy connecting to a cluster elsewhere to work.

**Comment by fasaxc** [comment]
Do we care about the case where you had one node in your cluster and add a new one (or if you're lazy and don't pass a full list to Felix)?  Maybe we should have an explicit EtcdIsProxy flag?

**Comment by TrimBiggs** [comment]
Fixed by #882
